scylla-macros/src/serialize/row.rs

```diff
@@ -206,47 +236,87 @@ impl Generator for ColumnSortingGenerator<'_> {
+        let nonflattened: Vec<_> = self
+            .ctx
+            .fields
+            .iter()
+            .filter(|f| !f.attrs.flatten)
+            .collect();
+        let nonflattened_fields: Vec<_> = nonflattened.iter().map(|f| &f.ident).collect();
+        let nonflattened_visited_flag_names: Vec<_> = nonflattened_fields
             .iter()
             .map(|ident| format_ident!("__visited_flag_{}", ident))
             .collect();
-        let num_fields = visited_flag_names.len();
+        let nonflattened_types: Vec<_> = nonflattened.iter().map(|f| &f.typ).collect();
 
         let partial_struct: syn::ItemStruct = parse_quote! {
             pub struct #partial_struct_name #partial_generics {
-                #(#fields: &#partial_lt #types,)*
-                #(#visited_flag_names: bool,)*
+                #(#nonflattened_fields: &#partial_lt #nonflattened_types,)*
+                #(#flattened_fields: <#flattened_types as #crate_path::SerializeRowByName>::Partial<#partial_lt>,)*
+                #(#flattened_visited_flag_names: bool,)*
+                #(#nonflattened_visited_flag_names: bool,)*
                 remaining_count: usize,
             }
         };
 
+        // 2. Implement serialization for the partial struct
+        let nonflattened_columns: Vec<_> = nonflattened.iter().map(|f| f.column_name()).collect();
+        let (impl_generics, ty_generics, where_clause) = self.ctx.generics.split_for_impl();
         let serialize_field_block: syn::Block = if self.ctx.fields.is_empty() {
             parse_quote! {{
                 ::std::result::Result::Ok(#crate_path::ser::row::FieldStatus::NotUsed)
             }}
         } else {
             parse_quote! {{
                 match spec.name() {
-                    #(#columns => {
+                    // first check if the spec name matches a non-flattened column
+                    #(#nonflattened_columns => {
                         #crate_path::ser::row::serialize_column::<#struct_name #ty_generics>(
-                            &self.#fields, spec, writer,
+                            &self.#nonflattened_fields, spec, writer,
                         )?;
-                        if !self.#visited_flag_names {
-                            self.#visited_flag_names = true;
-                            self.remaining_count -=1;
+                        if !self.#nonflattened_visited_flag_names {
+                            self.#nonflattened_visited_flag_names = true;
+                            self.remaining_count -= 1;
                         }
                     })*
-                    _ => {
+                    // if not, then check if any flattened field has a column for it
+                    _ => 'flatten_try: {
+                        #({
+                            match self.#flattened_fields.serialize_field(spec, writer)? {
+                                // there is a column and the field is done
```

Comment:
Ugh, I reall thought I would give approve this time, but there is one more unqualified call :(